xml_converter/src/xml_converter.cpp

```diff
@@ -130,50 +149,58 @@ void write_protobuf_file(string proto_filepath, map<string, Category>* marker_ca
-                if (category == proto_categories.end()) {
-                    cout << "Unknown MarkerCategory " << proto_pois.trail(i).category().name() << endl;
-                }
-                else {
-                    // This loop adds the name of the category and all of it's parents
-                    string name;
-                    for (unsigned int j = 0; j < split_categories.size(); j++) {
-                        name += split_categories[j];
-                        category_includes.insert(name);
-                        name += ".";
+            else if (parsed_poi->classname() == "Trail") {
+                Trail* trail = dynamic_cast<Trail*>(parsed_poi);
+                if (trail->map_id == map_id) {
+                    waypoint::Trail proto_trail = trail->as_protobuf();
+                    vector<string> split_categories = split(proto_trail.category().name(), ".");
+                    auto category = proto_categories.find(lowercase(split_categories[split_categories.size() - 1]));
+                    if (category == proto_categories.end()) {
+                        cout << "Unknown MarkerCategory " << proto_trail.category().name() << endl;
+                    }
+                    else {
+                        // This loop adds the name of the category and all of it's parents
+                        string name;
+                        for (unsigned int j = 0; j < split_categories.size(); j++) {
+                            name += split_categories[j];
+                            category_includes.insert(name);
+                            name += ".";
+                        }
                     }
+                    output_message.add_trail()->CopyFrom(proto_trail);
                 }
-                output_message.add_trail()->CopyFrom(proto_pois.trail(i));
             }
         }
-        // In XML, Marker_Categories has a tree hierarchy while POIS have a flat hierarchy.
+        // In the XML, Marker_Categories is a tree hierarchy while
+        // POIS have a flat hierarchy.
         // This is preserved in the protobuf for ease of translation.
-        // We are doing a removal instead of an insertion because each parent category contains the data for all of its children.
-        // It would be impractical to include all of the data from each category except for the children and then re-add the children.
-        // That would require coping every non-child attribute individually and iterating over all the children.
-        // This pruning method is slower but ensures that the all wanted information is kept.
+        // We are doing a removal instead of an insertion because each parent
+        // category contains the data for all of its children.
+        // It would be impractical to include all of the data from each category
+        // except for the children and then iterating over all the children.
+        // This pruning method is slower but ensures that information is kept.
```

Comment:
```suggestion
        // In the XML, MarkerCategories have a tree hierarchy while POIS have a
        // flat hierarchy. This is preserved in the protobuf for ease of
        // translation. We are doing a removal instead of an insertion because
        // each parent category contains the data for all of its children. It
        // would be impractical to include all of the data from each category
        // except for the children and then iterating over all the children.
        // This pruning method is slower but ensures that information is kept.
```